snowfall/models/transformer.py

```diff
@@ -385,6 +385,73 @@ def forward(self, x: Tensor) -> Tensor:
         return x
 
 
+class VggSubsampling(nn.Module):
+    """Trying to follow the setup described here https://arxiv.org/pdf/1910.09799.pdf
+       This paper is not 100% explicit so I am guessing to some extent,
+       and trying to compare with other VGG implementations.
+
+    Args:
+        idim: Input dimension.
+        odim: Output dimension.
+
+    """
+
+    def __init__(self, idim: int, odim: int) -> None:
+        """Construct a VggSubsampling object.   This uses 2 VGG blocks with 2
+           Conv2d layers each, subsampling its input by a factor of 4 in the
+           time dimensions.
+
+           Args:
+             idim:  Number of features at input, e.g. 40 or 80 for MFCC
+                    (will be treated as the image height).
+             odim:  Output dimension (number of features), e.g. 256
+        """
+        super(VggSubsampling, self).__init__()
+
+        cur_channels = 1
+        layers = []
+        block_dims = [32,64]
+
+        # The decision to use padding=1 for the 1st convolution, then padding=0
+        # for the 2nd and for the max-pooling, and ceil_mode=True, was driven by
+        # a back-compatibility concern so that the number of frames at the
+        # output would be equal to:
+        #  (((T-1)//2)-1)//2.
+        # We can consider changing this by using padding=1 on the 2nd convolution,
+        # so the num-frames at the output would be T//4.
```

Comment:
Would it make sense to add a method to each model that computes the subsampled shape from an original shape (or dim)? It’d be easier to support differing subsampling strategies.

Reply:
I think another option is to keep our models subsampling exactly with the value `subsampling_factor`,  like `T // 4` here. This would make the subsampling easier. We only need to pass the `subsampling_factor` to the corresponding function.

In this way, we'll need one side padding in some cases, like `Conv2dSubsampling` in our Transformer. The defalut convolution layers in PyTorch don't support one side padding. But `torch.nn.ConstantPad2d` could do this. We'll have to handle convolutions carefully in our model. Don't know if there will be any other side effects.

Reply:
I think something even easier would be to simply modify our code so it's insensitive to small differences in padding.   E.g. when computing the length of supervision segments we could just use T//4, but then truncate all segments so they fit within the actual computed output.  If this is possible?

Reply:
I'm not sure if I understood your idea correctly. If you mean truncating `num_frames` in supervision segments when you expect them to be longer than the nnet output, I think we'd have to be careful and truncate it from both sides if a segment begins in the middle of a cut (i.e. shift the `start_frame` too).

Reply:
I guess I mean we can get the num_frames of the cuts always by dividing by 4, but then just post-process them so they never go beyond the bounds of the actual nnet output, in case that is slightly less than the original T//4.  That is-- not have any specific requirements on how our nnet handles the downsampling, but just deal with it after the fact.

Reply:
I think the easiest way to resolve this would be to add an option with default:
   `allow_truncate=0`
to DenseFsaVec.__init__() in k2, so it would truncate the last few frames from a segment if necessary to
make it fit within the nnet output; we could set this to 2 or 3, for instance.  It seems to me that this
would be very easy to implement.  The only thing that would not be quite right is in mmi.py where we do:
```
        tot_score, tot_frames, all_frames = get_tot_objf_and_num_frames(
            tot_scores,
            supervision_segments[:, 2]
        )
```
supervision_segments[:, 2] may not have quite the right lengths.  But we could easily add a member function to the
DenseFsaVec object in Python, that would provide the correct length (e.g. we could save something in the constructor, to avoid an extra GPU->CPU transfer).

Reply:
.. does anyone have time to do this in k2?   (or let me know if there are any objections).

Reply:
ok, I am creating a PR in k2.